### ISSUE TITLE
initEvent('mdl-componentupgraded') causing an error in Firefox Update 45

### ISSUE THREAD

**Issue by DevJazz** [google/material-design-lite#4188]
**Version:** Material-Design-Lite 1.1.2.
**Function:** upgradeElementInternal
**Error:** NS_ERROR_UNEXPECTED: element.dispatchEvent(ev);

Hi all, this may need checking through with anyone else using Firefox 45.0 but today I noticed MDL stopped working on the following lines.

```
      var ev = document.createEvent('Events');
      ev.initEvent('mdl-componentupgraded', true, true);
      element.dispatchEvent(ev);
```

Commenting these lines out stops the error and I believe it's failing due to the following deprecation:
https://developer.mozilla.org/en-US/docs/Web/API/Event/initEvent

If this is merely an issue on my side apologies! Anyone else run into this issue?

**Comment by sgomes** [comment]
Thanks, @DevJazz, I had no idea we were using a deprecated call! I'll look into this.

**Comment by sgomes** [comment]
I have a fix prepared, but can't seem to reproduce the issue on getmdl.io. Is there any particular URL where you ran into this issue, @DevJazz?

**Comment by DevJazz** [comment]
@sgomes Thanks for the quick response! No specific URL as it's an internal development site but a majority of modules (including MDL) are loaded through RequireJS so not sure if it's something becoming out of sync or did not load correctly.

Never had an issue before but after recently updating to Firefox 45 today the error started occurring. Not sure if it's just luck it's been working beforehand or the update caused the issue! Sorry I can't give much more help than that!

**Comment by sgomes** [comment]
@DevJazz No worries, this is something we should fix anyway. I was just hoping to find a direct way of testing my fix :)

**Comment by DevJazz** [comment]
@sgomes Thank you for the fix :)

**Comment by Garbee** [comment]
I don't see where the "deprecation" should cause issues. Nothing is in the FF45 Release Notes about dropping this nor in their bugtracker as far as I can tell.

**Comment by sgomes** [comment]
I couldn't find anything either, and I can't reproduce the issue. Still, it seems like a sensible fix, although maybe not as urgent as I feared. Good material for our next point release.

**Comment by sgomes** [comment]
Good find, @Garbee! Is this a Firefox bug, or is it an instance of unspecified behaviour on the web platform?

**Comment by Garbee** [comment]
Not sure tbh. Considering it "was working" and is now "not working" I'd say FF bug, since other browsers it works fine.

It is deprecated anyways in favor of the `Event` object. So, I doubt the web platform would put any work into fixing any edge-cases that may exist.

**Comment by Garbee** [comment]
The custom event upgrade is worth doing to avoid this. So let's go for it. But, the bug isn't internal to us, it is an FF regression. Still urgent since MDL sites that hit this case are broken until they update to a new patch release with the fix in place.

**Comment by sgomes** [comment]
Alright, let's go with the fix in #4191 for now and see how that pans out. I'll do a point release next week with this and other fixes that we get merged until then.

**Comment by sgomes** [comment]
Merged.
